Lib/fractions.py

```diff
@@ -57,22 +57,42 @@ def _gcd(a, b):
 """, re.VERBOSE | re.IGNORECASE)
 
 
+def _as_integer_ratio(obj):
+    """
+    Return a pair (numerator, denominator) if obj has an ``as_integer_ratio``
+    method or is an instance of ``numbers.Rational``. Return ``NotImplemented``
+    if neither works.
+    """
+    try:
+        f = obj.as_integer_ratio
+    except AttributeError:
+        pass
+    else:
+        return f()
+
+    if isinstance(obj, numbers.Rational):
+        return (obj.numerator, obj.denominator)
+
+    return NotImplemented
```

Comment:
Is returning the ``NotImplemented`` constant appropriate in this case? Based on my impression from the [docs](https://docs.python.org/3/library/constants.html#NotImplemented), it looks like it should be used only by binary special methods.

Reply:
The docs indeed say that `NotImplemented` should be returned from such binary operators. However, I don't interpret it as saying that this is the *only* purpose of `NotImplemented`.

In any case, we need to do something when the operation is not supported. I could have raised an exception, but it seemed simpler and more efficient to define a special return value instead (especially given that this is an internal function). And when you're looking for a special return value to mean that an operation is not supported, `NotImplemented` is the most logical choice.

Reply:
I don't want to bother changing the documentation. I really find it a minor detail.

Reply:
There are actually multiple existing uses of `NotImplemented` on things which are not binary operators:

1. [`__subclasshook__`](https://docs.python.org/3/library/abc.html#abc.ABCMeta.__subclasshook__)
2. [`reducer_override`](https://docs.python.org/3.9/library/pickle.html#pickle.Pickler.reducer_override)
3. [`__length_hint__`](https://docs.python.org/3/reference/datamodel.html#object.__length_hint__) Note that returning `NotImplemented` is not documented but it is supported as special case.

Reply:
@jdemeyer 
> I don't want to bother changing the documentation. I really find it a minor detail.

For clarification, I'm not saying this is at all necessary as a part of this PR or that you would have to do it. It would be a completely separate issue involving a documentation change to provide more information about use cases for the ``NotImplemented`` constant.

The only reason that I mentioned it here is because this conversation gave me the idea. I think that providing an example of other use cases (besides binary special methods) for it would be helpful for users.

Thanks for the examples.

Reply:
If somebody wants to change the documentation of `NotImplemented`, I would support that. But I don't plan to do it myself.

Reply:
I'll open an issue for it on bpo and see what the others think.